### PR TITLE
1807/eb/web 3741 update all management page urls with each pagination view

### PR DESCRIPTION
**Pull request by scottpchow23**
Summary:

* Added `jquery-deparam` gem to pass tests.
* Adds logic for pagination which uses a parameter in the url to store/load the page.
* Light refactoring of reused methods in local_data_table_spec.js

**Note**: Test `"should set the max width to the DataTables_sort_interceptor on resize"` fails on the second time you run a test in `yarn run dev`, which seems to indicate a fragile test.

### CODE REVIEW

File: src/data_table/local_data_table.js [hunk 1]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
...
+      this.page(this._parseQueryString(window.location.search));
+    }
+  }
+
+  _createSearchString(pageNumber) {
+    var queryString = window.location.search;
+    var searchString = "?";
+    pageNumber++;
+    var pageString = "page=" + pageNumber;
+    if (queryString[0] === '?') {
```
Comment (jebentier) [inline src/data_table/local_data_table.js]: i think this can all be simplified to read a lot nicer.  its looks like all this is doing is trying to add the page number to the already existing query params.

```
let urlParams = $.deparam(window.location.href.split("?")[1] || "");
urlParams.page = pageNumber;
return "?" + $.param(urlParams)
```

File: src/data_table/local_data_table.js [hunk 2]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
...
+    let pageNumber = this._parseQueryString(window.location.search);
+    if (pageNumber >= 0) {
+      this.page(this._parseQueryString(window.location.search));
+    }
+  }
+
+  _createSearchString(pageNumber) {
+    var queryString = window.location.search;
+    var searchString = "?";
+    pageNumber++;
```
[the remark this answers is not in the document]
Comment (jebentier) [reply src/data_table/local_data_table.js]: i think it's better for this method to only be in charge of creating the search string, and make it the calling codes responsibility to pass in the appropriate pageNumber, removing the ++ here and doing `+ 1` to anywhere that is calling this and needs the page number incremented in the query string

File: src/data_table/local_data_table.js [hunk 3]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
...
+      searchString += pageString;
+    }
+    return searchString;
+  }
+
+  _parseQueryString(search) {
+    if (search[0] === '?') {
+      search = search.substr(1);
+    }
+    let parameters = $.deparam(search);
```
Comment (jebentier) [inline src/data_table/local_data_table.js]: looks like you're toggling back and forth between es5 and es6 in here, solidify in one please.

Comment (jebentier) [reply src/data_table/local_data_table.js]: this and the lines above it can be simplified to `let parameters = $.deparam(search.split('?')[1] || '')`

File: src/data_table/local_data_table.js [hunk 4]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
...
+      var parameters = $.deparam(queryString);
+      parameters.page = pageNumber;
+      searchString += $.param(parameters);
+    } else {
+      searchString += pageString;
+    }
+    return searchString;
+  }
+
+  _parseQueryString(search) {
```
Comment (jebentier) [inline src/data_table/local_data_table.js]: this method name isn't really descriptive of what it's doing, something more like `_parsePageNumberFromQueryString` is more descriptive.

File: src/data_table/local_data_table.js [hunk 5]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
...
+    if (search[0] === '?') {
+      search = search.substr(1);
+    }
+    let parameters = $.deparam(search);
+    let page = parseInt(parameters.page) - 1;
+    if (isNaN(page)) { return 0; }
+    return parseInt(parameters.page) - 1;
```
Comment (jebentier) [inline src/data_table/local_data_table.js]: for more clarity to the reader, i would put this into a full `if/else` so it's clear that NaN is a special case

Comment (nburwell) [reply src/data_table/local_data_table.js]: also this last line is just `page` right?  you already parsed and subtracted in L509.

```
let page = parseInt(parameters.page) - 1;
if (isNaN(page)) {
  return 0;
} else {
  return page;
}
```

File: src/data_table/local_data_table.js [hunk 6]
```diff
@@ -459,6 +461,46 @@ class LocalDataTable extends View {
...
+
+  _pageToSearchPage() {
+    let pageNumber = this._parsePageNumberFromQueryString();
+    if (pageNumber >= 0) {
+      this.page(pageNumber);
+    }
+  }
+
+  _createSearchString(pageNumber) {
+    let urlParameters = $.deparam(window.location.href.split("?")[1] || "");
```
Comment (jebentier) [inline src/data_table/local_data_table.js]: this line is duplicated, you can make a helper method for urlParameters

File: src/data_table/local_data_table.js [hunk 7]
```diff
@@ -491,6 +539,9 @@ class LocalDataTable extends View {
       aoColumns: this._columnManager.dataTableColumnsConfig(),
       aaSorting: this._columnManager.dataTableSortingConfig(),
       fnDrawCallback: this._onDraw,
+      iDisplayStart: displayStart,
+      iRecordsTotal: recordTotal,
+      iRecordsDisplay: recordTotal,
```
Comment (nburwell) [inline src/data_table/local_data_table.js]: can you move these up next to `iDsplayLength` as they are related?

File: src/data_table/local_data_table.js [hunk 8]
```diff
@@ -480,6 +526,8 @@ class LocalDataTable extends View {
   }
 
   _dataTableConfig() {
+    let displayStart = this._parsePageNumberFromQueryString() * this.paginateLength;
+    let recordTotal = displayStart + 10;
```
Comment (nburwell) [inline src/data_table/local_data_table.js]: this 10 needs to be `this.paginateLength` instead

File: src/data_table/local_data_table.js [hunk 9]
```diff
@@ -459,6 +461,56 @@ class LocalDataTable extends View {
     this.dataTable.on("page", this._bulkCheckboxAdjust);
   }
 
+  _setupPaginationHistory() {
+    this.dataTable.on("page", function() {
+      var page = this.dataTable.fnPagingInfo().iPage;
+      if (page !== this._parseQueryString(window.location.search)) {
+        history.pushState({}, "pagination", this._createSearchString(this.dataTable.fnPagingInfo().iPage));
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply src/data_table/local_data_table.js]: ✅

File: spec/data_table/local_data_table_spec.js [hunk 10]
```diff
@@ -145,6 +153,64 @@ describe("DataTable Plugin", function() {
...
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/1 to 10/);
+      });
+      it("should load into page 1 if page parameter is out of bounds", function() {
+        history.pushState({}, "pagination", "?page=500");
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/1 to 10/);
+        expect(window.location.search).toMatch(/page=1/);
+      });
+      it("shouldn't get tripped up by other query variables in the url", function() {
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/local_data_table_spec.js]: ✅

File: spec/data_table/server_side_data_table_spec.js [hunk 11]
```diff
@@ -203,6 +204,74 @@ describe("DataTable Plugin", function() {
     jasmine.Ajax.uninstall();
   });
 
+  describe("server side pagination history", function() {
+    afterEach(function() {
+      history.pushState({}, "pagination", "?");
+    });
+
+    it("should handling going back", function() {
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/server_side_data_table_spec.js]: ✅

File: spec/data_table/server_side_data_table_spec.js [hunk 12]
```diff
@@ -203,6 +204,74 @@ describe("DataTable Plugin", function() {
...
+      const table = new TestDataTable({ collection: this.collection });
+      table.render();
+      jasmine.Ajax.requests.mostRecent().response(this.mockResponse.get());
+      table.page("next");
+      jasmine.Ajax.requests.mostRecent().response(this.mockResponse.get());
+      table.page("next");
+      jasmine.Ajax.requests.mostRecent().response(this.mockResponse.get());
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/server_side_data_table_spec.js]: ✅

File: src/data_table/local_data_table.js [hunk 13]
```diff
@@ -459,6 +461,50 @@ class LocalDataTable extends View {
...
+    this.dataTable.on("page", () => {
+      let page = this.dataTable.fnPagingInfo().iPage;
+      if (page !== this._parsePageNumberFromQueryString()) {
+        history.pushState({}, "pagination", this._createSearchString(this.dataTable.fnPagingInfo().iPage + 1));
+      }
+    });
+    window.onpopstate = () => {
+      this._pageToSearchPage();
+    };
+    this.on("render", this._afterRender());
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply src/data_table/local_data_table.js]: I think I like the second idea more because listeners are hard to track down. ✅

File: spec/data_table/local_data_table_spec.js [hunk 14]
```diff
@@ -145,6 +153,70 @@ describe("DataTable Plugin", function() {
...
+        history.pushState({}, "pagination", "?page=500");
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/1 to 10/);
+        expect(window.location.search).toMatch(/page=1/);
+      });
+
+      it("shouldn't get tripped up by other query variables in the url", function() {
+        history.pushState({}, "pagination", "?something=awesome&page=3");
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/21 to 30/);
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/local_data_table_spec.js]: ✅

File: spec/data_table/server_side_data_table_spec.js [hunk 15]
```diff
@@ -203,6 +204,68 @@ describe("DataTable Plugin", function() {
...
+    });
+    it("should load into empty table if page parameter is out of bounds", function() {
+      history.pushState({}, "pagination", "?page=500");
+      const table = new TestDataTable({ collection: this.collection });
+      table.render();
+      jasmine.Ajax.requests.mostRecent().response(this.mockResponse.get());
+      expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/4,991 to 100/);
+      expect(window.location.search).toMatch(/page=500/);
+    });
+    it("shouldn't get tripped up by other query variables in the url", function() {
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/server_side_data_table_spec.js]: ✅

File: src/data_table/local_data_table.js [hunk 16]
```diff
@@ -459,6 +461,50 @@ class LocalDataTable extends View {
...
+    if (pageNumber >= 0) {
+      this.page(pageNumber);
+    }
+  }
+
+  _urlParameters() {
+    return $.deparam(window.location.href.split("?")[1] || "");
+  }
+
+  _createSearchString(pageNumber) {
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply src/data_table/local_data_table.js]: ✅

File: src/data_table/local_data_table.js (hunk 7 above)
Comment (scottpchow23) [reply src/data_table/local_data_table.js]: ✅

File: src/data_table/local_data_table.js (hunk 8 above)
Comment (scottpchow23) [reply src/data_table/local_data_table.js]: ✅

File: spec/data_table/local_data_table_spec.js [hunk 17]
```diff
@@ -145,6 +153,64 @@ describe("DataTable Plugin", function() {
...
+        history.pushState({}, "pagination", "?");
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/1 to 10/);
+      });
+      it("should load into page 1 if page parameter is out of bounds", function() {
+        history.pushState({}, "pagination", "?page=500");
+        table.render();
+        expect(table.$el.find('.dataTables_info')[0].innerText).toMatch(/1 to 10/);
+        expect(window.location.search).toMatch(/page=1/);
+      });
```
[the remark this answers is not in the document]
Comment (scottpchow23) [reply spec/data_table/local_data_table_spec.js]: ✅

File: src/data_table/local_data_table.js [hunk 18]
```diff
@@ -77,6 +77,7 @@ class LocalDataTable extends View {
     this._triggerChangeSelection();
     this.paginate && this._setupPaginationHistory();
     this.trigger("render");
+    this.paginate && this._afterRender();
```
Comment (nburwell) [inline src/data_table/local_data_table.js]: as discussed let's move this paginate check into the afterRender method

File: spec/data_table_plugin_spec.js [hunk 19]
```diff
@@ -1272,6 +1272,7 @@ describe("DataTable Plugin", function() {
         stubResizeMode(table);
         triggerMouseMoveEvent(table, { mouseX: -10 });
         const endWidth = startWidth - 10;
+        //TODO: Fix this test so it doesn't break after the first time in yarn run dev
```
[the remark this answers is not in the document]
Comment (rjkearns) [reply spec/data_table_plugin_spec.js]: Was this TODO ever fixed?

File: vendor/jquery.dataTables-1.9.4.js [hunk 20]
```diff
@@ -6487,6 +6487,9 @@
 			_fnMap( oSettings, oInit, "oSearch", "oPreviousSearch" );
 			_fnMap( oSettings, oInit, "aoSearchCols", "aoPreSearchCols" );
 			_fnMap( oSettings, oInit, "iDisplayLength", "_iDisplayLength" );
+      _fnMap( oSettings, oInit, "iDisplayStart", "_iDisplayStart" );
+      _fnMap( oSettings, oInit, "iRecordsTotal", "_iRecordsTotal" );
+      _fnMap( oSettings, oInit, "iRecordsDisplay", "_iRecordsDisplay" );
```
Comment (rjkearns) [inline vendor/jquery.dataTables-1.9.4.js]: nit: spacing is off here

File: src/data_table/local_data_table.js [hunk 21]
```diff
@@ -548,6 +550,15 @@ class LocalDataTable extends View {
     };
   }
 
+  _getSafeDisplayStartFromPageNumber() {
+    let pageIndex = this._parsePageNumberFromQueryString() - 1;
+    if (pageIndex < 0) {
+      return 0;
+    } else {
+      return (this._parsePageNumberFromQueryString() - 1) * this.paginateLength;
```
Comment (rjkearns) [inline src/data_table/local_data_table.js]: `return pageIndex * this.paginateLength;`

File: vendor/jquery.dataTables-1.9.4.js (hunk 20 above)
Comment (scottpchow23) [reply vendor/jquery.dataTables-1.9.4.js]: Weird, it's not showing up in RubyMine as having weird spacing.

File: dist/backdraft.js [hunk 22]
```diff
@@ -0,0 +1,6448 @@
+var Backdraft =
```
Comment (rjkearns) [inline dist/backdraft.js]: is this file needed?
